src/composables/approvals/useTokenApprovalActions.ts

```diff
@@ -198,29 +201,64 @@ export default function useTokenApprovalActions() {
-    return approvalsRequired.map(amountToApprove => {
-      const token = getToken(amountToApprove.address);
-
-      return {
-        label: actionLabel(actionType, token.symbol),
-        loadingLabel: t('investment.preview.loadingLabel.approval'),
-        confirmingLabel: t('confirming'),
-        stepTooltip: actionTooltip(actionType, token.symbol),
-        action: () =>
-          approveToken({
-            token,
-            normalizedAmount: amountToApprove.amount,
-            spender,
-            actionType,
-            forceMax,
-          }),
-        postActionValidation: () => isApprovalValid(amountToApprove, spender),
-        actionInvalidReason: {
-          title: t('actionSteps.approve.invalidReason.title'),
-          description: t('actionSteps.approve.invalidReason.description'),
-        },
-      };
-    });
+    return flatten(
+      approvalsRequired.map(amountToApprove => {
+        const token = getToken(amountToApprove.address);
+        const actions: TransactionActionInfo[] = [];
+        /**
+         * Some tokens require setting approved amount to 0 before changing the
+         * approval amount. This injects another action to do that.
+         */
+        if (
+          configService.network.tokens.DoubleApprovalRequired?.includes(
+            token.address
+          ) &&
+          allowanceFor(token.address, spender).gt(0)
+        ) {
+          actions.push({
+            label: actionLabel(actionType, token.symbol),
+            loadingLabel: t('investment.preview.loadingLabel.approval'),
+            confirmingLabel: t('confirming'),
+            stepTooltip: t('transactionSummary.tooltips.unapprove', [
+              token.symbol,
+            ]),
+            action: () =>
+              approveToken({
+                token,
+                normalizedAmount: '0',
+                spender,
+                actionType,
+                forceMax,
+              }),
+            actionInvalidReason: {
+              title: t('actionSteps.approve.invalidReason.title'),
+              description: t('actionSteps.approve.invalidReason.description'),
+            },
+          });
+        }
+
+        actions.push({
+          label: actionLabel(actionType, token.symbol),
+          loadingLabel: t('investment.preview.loadingLabel.approval'),
+          confirmingLabel: t('confirming'),
+          stepTooltip: actionTooltip(actionType, token.symbol),
+          action: () =>
+            approveToken({
+              token,
+              normalizedAmount: amountToApprove.amount,
+              spender,
+              actionType,
+              forceMax,
+            }),
+          postActionValidation: () => isApprovalValid(amountToApprove, spender),
+          actionInvalidReason: {
+            title: t('actionSteps.approve.invalidReason.title'),
+            description: t('actionSteps.approve.invalidReason.description'),
+          },
+        });
+        return actions;
+      })
+    );
```

Comment:
Could you clean this up so it's easier to parse, something like:
```suggestion
    return flatten(
      approvalsRequired.map(amountToApprove => {
        const token = getToken(amountToApprove.address);
        const actions: TransactionActionInfo[] = [];

        if (isDoubleApprovalRequired(token.address, spender)) {
          actions.push(zeroApprovalAction(...params));
        }

        actions.push(approvalAction(...params));
     
        return actions;
      })
    );
```
You'll need to extract the action construction into function etc.

Reply:
Fixed